llama/src/main/java/com/cloudera/llama/am/yarn/YarnRMConnector.java

```diff
@@ -806,10 +815,51 @@ public void onContainersAllocated(List<Container> containers) {
           LOG.trace("Reservation resource '{}' removed from YARN", resource);
 
           queue(new ContainerHandler(ugi, resource, container, Action.START));
+
+          /*Remove the granted request from anyLocationResourceIdToRequestMap if it is there*/
+          anyLocationResourceIdToRequestMap.remove(resource.getResourceId());
         }
       } else {
-        LOG.error("No matching request for {}. Releasing the container.",
+        LOG.debug("No strong request match for {}. Adding to the list of unclaimed containers.",
             container);
+        unclaimedContainers.add(container);
+      }
+    }
+    /*Matching YARN resources against requests relaxing locality*/
+    /*Doing this in the separate loop as strong match should be preferred */
+    for (Container container : unclaimedContainers) {
+      /*Find pending request that relax locality which can get use of unclaimed containers*/
+      boolean containerIsClaimed = false;
+      Iterator<Map.Entry<UUID, LlamaContainerRequest>> iterator = anyLocationResourceIdToRequestMap.entrySet().iterator();
+      while (iterator.hasNext()) {
+        Map.Entry<UUID, LlamaContainerRequest> entry = iterator.next();
+        LlamaContainerRequest request = entry.getValue();
+        /*Matching by the capacity only*/
+        if(request.getResourceAsk().getCpuVCoresAsk() == container.getResource().getVirtualCores() &&
+                request.getResourceAsk().getMemoryMbsAsk() == container.getResource().getMemory()) {
+
+          RMResource resource = request.getResourceAsk();
```

Comment:
This code is duplicated from the case where there is a strong match. Move it to a private helper method and reuse at both places?